accumulator/utils.go

```diff
@@ -189,6 +190,127 @@ func treeRows(n uint64) (e uint8) {
 	return
 }
 
+// given n leaves, how deep is the tree?
+// iterate shifting left until greater than n
+func treeRows2(n uint64) uint8 {
+	// treeRows works by:
+	// 1. find the next power of 2 from the given n leaves.
+	//
+	// For example, if the given number is 9, the next power of 2 is
+	// 16. This log2 of this number is how many rows there are in the
+	// given tree.
+
+	// Find the next power of 2
+	n--
+	n |= n >> 1
+	n |= n >> 2
+	n |= n >> 4
+	n |= n >> 8
+	n |= n >> 16
+	n |= n >> 32
+	n++
+
+	// log of 2 is the tree depth/height
+	return uint8(math.Log2(float64(n)))
+}
+
+// given n leaves, how deep is the tree?
+// iterate shifting left until greater than n
+func treeRows3(n uint64) uint8 {
+	// treeRows works by:
+	// 1. find the next power of 2 from the given n leaves.
+	//
+	// For example, if the given number is 9, the next power of 2 is
+	// 16. This log2 of this number is how many rows there are in the
+	// given tree.
+
+	// Find the next power of 2
+	n--
+	n |= n >> 1
+	n |= n >> 2
+	n |= n >> 4
+	n |= n >> 8
+	n |= n >> 16
+	n |= n >> 32
+	n++
+
+	r := uint8(0)
+	x := uint8(0)
+
+	// Find log2 of the number
+	if (n & 0xAAAAAAAAAAAAAAAA) != 0 {
+		r = 1
+	}
+
+	if (n & 0xFFFFFFFF00000000) != 0 {
+		x = 1 << 5
+		r |= x
+	}
+
+	if (n & 0xFFFF0000FFFF0000) != 0 {
+		x = 1 << 4
+		r |= x
+	}
+
+	if (n & 0xFF00FF00FF00FF00) != 0 {
+		x = 1 << 3
+		r |= x
+	}
+
+	if (n & 0xF0F0F0F0F0F0F0F0) != 0 {
+		x = 1 << 2
+		r |= x
+	}
+
+	if (n & 0xCCCCCCCCCCCCCCCC) != 0 {
+		x = 1 << 1
+		r |= x
+	}
+
+	return r
+}
+
+// given n leaves, how deep is the tree?
+// iterate shifting left until greater than n
+func treeRows4(n uint64) uint8 {
+	// treeRows works by:
+	// 1. find the next power of 2 from the given n leaves.
+	//
+	// For example, if the given number is 9, the next power of 2 is
+	// 16. This log2 of this number is how many rows there are in the
+	// given tree.
+
+	// Find the next power of 2
+	n--
+	n |= n >> 1
+	n |= n >> 2
+	n |= n >> 4
+	n |= n >> 8
+	n |= n >> 16
+	n |= n >> 32
+	n++
+
+	// Find log2 of the number
```

Comment:
could also use `bits.TrailingZeros` to find the log2 of n. looks less cryptic and it's also optimized.

Reply:
wrote a benchmark `Benchmark5TreeRows` with it and i seems to be even faster:
```
~/go/src/github.com/mit-dci/utreexo/accumulator(treerow-calvin*) » go test -run=XXX -bench=Benchmark4TreeRows                                                             niklas@Niklass-MacBook-Pro
goos: darwin
goarch: amd64
pkg: github.com/mit-dci/utreexo/accumulator
Benchmark4TreeRows_HunThou-4    1000000000               0.00189 ns/op
Benchmark4TreeRows_Mil-4        1000000000               0.0143 ns/op
Benchmark4TreeRows_Bil-4        1000000000               0.143 ns/op
Benchmark4TreeRows_Tril-4              1        1432338272 ns/op
PASS
ok      github.com/mit-dci/utreexo/accumulator  3.371s
------------------------------------------------------------
~/go/src/github.com/mit-dci/utreexo/accumulator(treerow-calvin*) » go test -run=XXX -bench=Benchmark3TreeRows                                                             niklas@Niklass-MacBook-Pro
goos: darwin
goarch: amd64
pkg: github.com/mit-dci/utreexo/accumulator
Benchmark3TreeRows_HunThou-4    1000000000               0.000555 ns/op
Benchmark3TreeRows_Mil-4        1000000000               0.00665 ns/op
Benchmark3TreeRows_Bil-4        1000000000               0.0507 ns/op
Benchmark3TreeRows_Tril-4       1000000000               0.508 ns/op
PASS
ok      github.com/mit-dci/utreexo/accumulator  14.371s
------------------------------------------------------------
~/go/src/github.com/mit-dci/utreexo/accumulator(treerow-calvin*) » go test -run=XXX -bench=Benchmark2TreeRows                                                             niklas@Niklass-MacBook-Pro
goos: darwin
goarch: amd64
pkg: github.com/mit-dci/utreexo/accumulator
Benchmark2TreeRows_HunThou-4    1000000000               0.00199 ns/op
Benchmark2TreeRows_Mil-4        1000000000               0.0131 ns/op
Benchmark2TreeRows_Bil-4        1000000000               0.130 ns/op
Benchmark2TreeRows_Tril-4              1        1293551434 ns/op
PASS
ok      github.com/mit-dci/utreexo/accumulator  2.922s
------------------------------------------------------------
~/go/src/github.com/mit-dci/utreexo/accumulator(treerow-calvin*) » go test -run=XXX -bench=Benchmark5TreeRows                                                             niklas@Niklass-MacBook-Pro
goos: darwin
goarch: amd64
pkg: github.com/mit-dci/utreexo/accumulator
Benchmark5TreeRows_HunThou-4    1000000000               0.000037 ns/op
Benchmark5TreeRows_Mil-4        1000000000               0.000365 ns/op
Benchmark5TreeRows_Bil-4        1000000000               0.00328 ns/op
Benchmark5TreeRows_Tril-4       1000000000               0.0328 ns/op
PASS
ok      github.com/mit-dci/utreexo/accumulator  0.353s
```

Reply:
huh oh cool! Could you make a PR to my branch?

Reply:
sure, you should see the PR on your fork now.